pkg/resources/ali/ali.go

```diff
@@ -0,0 +1,691 @@
+	}
+
+	listFunctions := []aliListFn{
+		d.ListLoadBalancer,
+		d.ListRam,
+		d.ListSecurityGroup,
+		d.ListSSHKey,
+		d.ListVPC,
+		d.ListVolume,
+	}
+
+	for _, fn := range listFunctions {
+		resourceTrackers, err := fn()
+		if err != nil {
+			return nil, err
+		}
+		for _, t := range resourceTrackers {
+			resources[t.Type+":"+t.ID] = t
+		}
+	}
+
+	for k, t := range resources {
+		if t.Done {
+			delete(resources, k)
+		}
+	}
+
+	return resources, nil
+}
+
+func (d *clusterDiscoveryALI) ListScalingGroups() ([]*resources.Resource, error) {
+	var resourceTrackers []*resources.Resource
+
+	// ScalingGroup Name defined in /pkg/model/alimodel/context.go
+	// The should be changed at the same time.
+	remove := []string{
+		"masters." + d.clusterName,
+		"nodes." + d.clusterName,
+		"bastions." + d.clusterName,
+	}
+	pageNumber := 1
+	pageSize := 50
+
+	for {
+		describeScalingGroupsArgs := &ess.DescribeScalingGroupsArgs{
+			RegionId: common.Region(d.aliCloud.Region()),
+			Pagination: common.Pagination{
+				PageNumber: pageNumber,
+				PageSize:   pageSize,
+			},
+		}
+		groups, _, err := d.aliCloud.EssClient().DescribeScalingGroups(describeScalingGroupsArgs)
+
+		if err != nil {
+			return nil, fmt.Errorf("error listing ScalingGroup: %v", err)
+		}
+
+		scalingGroups := []ess.ScalingGroupItemType{}
+
+		for _, group := range groups {
+			for _, r := range remove {
+				if strings.Contains(group.ScalingGroupName, r) {
+					scalingGroups = append(scalingGroups, group)
+				}
+			}
+		}
+
+		for _, scalingGroup := range scalingGroups {
+
+			resourceTracker := &resources.Resource{
+				Name:    scalingGroup.ScalingGroupName,
+				ID:      scalingGroup.ScalingGroupId,
+				Type:    typeScalingGroup,
+				Deleter: DeleteScalingGroup,
+			}
+
+			d.scalingGroup[scalingGroup.ScalingGroupName] = scalingGroup.ScalingGroupId
+			resourceTrackers = append(resourceTrackers, resourceTracker)
+		}
+
+		if len(groups) < pageSize {
+			break
+		} else {
+			pageNumber++
+		}
+	}
+
+	return resourceTrackers, nil
+}
+
+func DeleteScalingGroup(cloud fi.Cloud, r *resources.Resource) error {
+	c := cloud.(aliup.ALICloud)
+	glog.V(2).Infof("Removing ScalingGroup with Id %s", r.ID)
+
+	// Force to delete the ScalingGroup
+	// TODO: Should we delete the group softly? Like set ForceDelete to false.
+	// That will be safer.
+
+	// All the resource of the ScalingGroup will be deleted.
+	deleteScalingGroupArgs := &ess.DeleteScalingGroupArgs{
+		ScalingGroupId: r.ID,
+		ForceDelete:    true,
+	}
+
+	_, err := c.EssClient().DeleteScalingGroup(deleteScalingGroupArgs)
+	if err != nil {
+		return fmt.Errorf("error deleting ScalingGroup: %v", err)
+	}
+
+	return nil
+}
+
+func (d *clusterDiscoveryALI) ListLoadBalancer() ([]*resources.Resource, error) {
+	var resourceTrackers []*resources.Resource
+
+	// LoadBalancer has the cluster tag: KubernetesCluster:$ClusterName
+	// Find the cluster loadBalancer by this key value pair.
+	// TODO: Should we check the tags of loadbalancer?
```

Comment:
AWS doesn't support tags on everything - if you can use tags, I think it's much more reliable ( & easier for users to understand ) :-)